src/java.base/share/classes/sun/security/ssl/DHasKEM.java

```diff
@@ -0,0 +1,354 @@
+                "ECDH", "EC", CurveDB.P_384),
+
+        P521(66, 2 * 66 + 1,
+                "ECDH", "EC", CurveDB.P_521),
+
+        X25519(32, 32,
+                "XDH", "XDH", NamedParameterSpec.X25519),
+
+        X448(56, 56,
+                "XDH", "XDH", NamedParameterSpec.X448),
+        ;
+        private final int secretLen;
+        private final int publicKeyLen;
+        private final String kaAlgorithm;
+        private final String keyAlgorithm;
+        private final AlgorithmParameterSpec spec;
+
+
+        Params(int secretLen, int publicKeyLen, String kaAlgorithm,
+                String keyAlgorithm, AlgorithmParameterSpec spec) {
+            this.spec = spec;
+            this.secretLen = secretLen;
+            this.publicKeyLen = publicKeyLen;
+            this.kaAlgorithm = kaAlgorithm;
+            this.keyAlgorithm = keyAlgorithm;
+        }
+
+        private boolean isEC() {
+            return this == P256 || this == P384 || this == P521;
+        }
+
+        private KeyPair generateKeyPair(SecureRandom sr) {
+            try {
+                KeyPairGenerator g = KeyPairGenerator.getInstance(keyAlgorithm);
+                g.initialize(spec, sr);
+                return g.generateKeyPair();
+            } catch (Exception e) {
+                throw new ProviderException("internal error", e);
+            }
+        }
+
+        private byte[] SerializePublicKey(PublicKey k) {
+            if (isEC()) {
+                ECPoint w = ((ECPublicKey) k).getW();
+                return ECUtil.encodePoint(w, ((NamedCurve) spec).getCurve());
+            } else {
+                byte[] uArray = ((XECPublicKey) k).getU().toByteArray();
+                ArrayUtil.reverse(uArray);
+                return Arrays.copyOf(uArray, publicKeyLen);
+            }
+        }
+
+        private PublicKey DeserializePublicKey(byte[] data) throws
+                IOException, NoSuchAlgorithmException,
+                InvalidKeySpecException {
+            KeySpec keySpec;
+            if (isEC()) {
+                NamedCurve curve = (NamedCurve) this.spec;
+                keySpec = new ECPublicKeySpec(
+                        ECUtil.decodePoint(data, curve.getCurve()), curve);
+            } else {
+                data = data.clone();
+                ArrayUtil.reverse(data);
+                keySpec = new XECPublicKeySpec(
+                        this.spec, new BigInteger(1, data));
+            }
+            return KeyFactory.getInstance(keyAlgorithm).
+                    generatePublic(keySpec);
+        }
+
+        private SecretKey DH(String alg, PrivateKey skE, PublicKey pkR)
+                throws NoSuchAlgorithmException, InvalidKeyException {
+            KeyAgreement ka = KeyAgreement.getInstance(kaAlgorithm);
+            ka.init(skE);
+            ka.doPhase(pkR, true);
+            return ka.generateSecret(alg);
+        }
+    }
+
+    private static class HybridService extends Provider.Service {
```

Comment:
Reminder: this would also be moved to the new Provider file....

Reply:
Moved to new HybridProvider file.